✨ Added any & all matching to {{#has}} helper

ErisDS opened the pull request:
This is the last piece of #8901. First commit is a refactor as the code had gotten a bit ridic!

closes #8901

- Adds support for

```
{{#has any="twitter, facebook, website"}}
{{#has any="author.facebook, author.twitter,author.website"}}
{{#has any="@blog.facebook, @blog.twitter, @labs.subscribers"}}
{{#has all="@labs.subscribers,@labs.publicAPI"}}
```

ErisDS commented:
Note: there's definitely a TODO here to change author and tag to work differently - the weird lowercase no-special char matching can be improved, but that could introduce backwards incompatible changes, so need to add a bunch more tests first.

kevinansfield commented:
@ErisDS I'm guessing by the naming these should be rebased and merged as separate commits?

ErisDS commented:
@kevinansfield I am happy either way - I like to keep refactors separate in commits so that you can more easily review what I've changed.

I am using the no-emoji on refactors.

The PR message and 2nd commit message are the same (I copy and pasted them from the slug PR and they were wrong, I just updated them haha).

Therefore, it doesn't matter if you squash and merge or rebase and merge, the changelog should look the same? Merely a matter of taste as to whether it's useful to keep the refactor in history.
